MoreLinq/Memoize.cs

```diff
@@ -0,0 +1,136 @@
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2016 Leandro F. Vieira (leandromoh). All rights reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Returns a <see cref="MemoizedEnumerable{T}"/> that lazily creates an in-memory
+        /// cache of the enumeration on first iteration, if it is not already an
+        /// in-memory source.
+        /// </summary>
+        /// <param name="source">The source sequence.</param>
+        /// <returns>A <see cref="MemoizedEnumerable{T}"/>.</returns>
+        public static IBufferedEnumerable<T> Memoize<T>(this IEnumerable<T> source)
+        {
+            return source.Memoize(false);
+        }
+
+        /// <summary>
+        /// Returns a <see cref="MemoizedEnumerable{T}"/> that lazily creates an in-memory
+        /// cache of the enumeration on first iteration, if it is not already an
+        /// in-memory source.
+        /// An additional argument specifies if buffering must happen even if the source implements
+        /// <see cref="ICollection{T}"/> (otherwise it is assumed that the source already serves as 
+        /// a buffer adequately).
+        /// </summary>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="forceBuffering">Force buffering, even if source is an <see cref="ICollection{T}" />
+        /// (otherwise source is assumed to already be in-memory and enumerated quickly, and hence that
+        /// buffering again would be a waste).</param>
+        /// <returns>A <see cref="MemoizedEnumerable{T}"/>.</returns>
+        public static IBufferedEnumerable<T> Memoize<T>(this IEnumerable<T> source, bool forceBuffering)
+        {
+            if (source == null) throw new ArgumentNullException("source");
+
+            return (source as MemoizedEnumerable<T>) ?? new MemoizedEnumerable<T>(source, forceBuffering);
+        }
+    }
+
+    /// <summary>
+    /// An enumeration that, when being iterated, lazily iterates through the source enumeration 
+    /// and cache the elements already iterated for future iterations. It supports partial iterations.
+    /// </summary>
+    /// <typeparam name="T">Type of the source sequence</typeparam>
+    public interface IBufferedEnumerable<T> : IEnumerable<T>
```

Comment:
To know that an `IEnumerable<T>` implements` ISelectEnumerable<T>` or `IWhereEnumerable<T> `dont help you with nothing, however if you know that an `IEnumerable<T>` implements an `IBufferedEnumerable<T>` you know that its elements are being cached and so you dont need to use ToList/ToArray or the like (as many people do), you end up forcing your code to wait until all the results have been pulled into memory from the server. 

To know that your `IEnumerable<T>` implements an `IBufferedEnumerable<T>` improves this situation by allowing you to start pulling from the `IEnumerable<T>` immediately, but without having to incur the penalty if you end up needing to iterate over it multiple times (as the data gets cached as it's been pulled from the source).

I also created this interface because the attached patch in the issue #87 has it. But if @atifaziz  tell us that this interface is no more needed in #87 or that it dont make sense anymore I can remove it without problems.